python/pyspark/ml/functions.py

```diff
@@ -106,6 +117,474 @@ def array_to_vector(col: Column) -> Column:
+                            ]
+                        else:
+                            raise ValueError("input_tensor_shapes must match columns")
+
+                    # run model prediction function on transformed (numpy) inputs
+                    preds = predict_fn(*multi_inputs)
+                elif num_expected == 1:
+                    # multiple input columns for single expected input
+                    if has_tensors:
+                        # tensor columns
+                        if len(batch.columns) == 1:
+                            # one tensor column and one expected input, vstack rows
+                            single_input = np.vstack(batch.iloc[:, 0])
+                        else:
+                            raise ValueError(
+                                "Multiple input columns found, but model expected a single "
+                                "input, use `struct` or `array` to combine columns into tensors."
+                            )
+                    else:
+                        # scalar columns
+                        if len(batch.columns) == 1:
+                            # single scalar column, remove extra dim
+                            single_input = np.squeeze(batch.to_numpy())
+                        elif not has_tuple:
+                            # columns grouped via struct/array, convert to single tensor
+                            single_input = batch.to_numpy()
+                        else:
+                            raise ValueError(
+                                "Multiple input columns found, but model expected a single "
+                                "input, use `struct` or `array` to combine columns into tensors."
+                            )
+
+                    # if input_tensor_shapes provided, try to reshape input
+                    if input_shapes:
+                        if len(input_shapes) == 1:
+                            single_input = single_input.reshape(
+                                [-1] + input_shapes[0]  # type: ignore
+                            )
+                        else:
+                            raise ValueError(
+                                "Multiple input_tensor_shapes found, but model expected one input"
+                            )
+
+                    # run model prediction function on transformed (numpy) inputs
+                    preds = predict_fn(single_input)
+                else:
+                    msg = "Model expected {} inputs, but received {} columns"
+                    raise ValueError(msg.format(num_expected, num_actual))
+
+                # return predictions to Spark
+                if isinstance(return_type, StructType):
+                    yield pd.DataFrame(preds)
+                elif isinstance(return_type, ArrayType):
+                    yield pd.Series(list(preds))  # type: ignore[misc]
+                else:
+                    yield pd.Series(np.squeeze(preds))  # type: ignore
```

Comment:
Issues:

* Suppose the model predict function returns multi-dimension array, the logic here causes the dimension information lost. How to return the result dimension information ?
* `yield pd.Series(np.squeeze(preds))` this might cause the UDF output rows not 1:1 map to input rows, which breaks spark scalar UDF contract.
* The UDF return type might be either `Iterator[pd.DataFrame]` or `Iterator[pd.Series]`, please update the type hint in `def predict`

Reply:
If `predict_fn` returns a 2 dimension array, and the spark UDF returned column is a spark array type, we need to convert the 2 dimension array into a one-dim array contains elements of array type.

Reply:
For the UDF return type, I tried using `Iterator[Union[pd.Series, pd.DataFrame]]`, but got this error from `pyspark/sql/pandas/typehints.py:135` during testing:
`NotImplementedError: Unsupported signature: (data: 'Iterator[Union[pd.Series, pd.DataFrame]]') -> 'Iterator[Union[pd.Series, pd.DataFrame]]'.`

Not sure if there's a solution to this, since the pandas_udf signatures seems to only support returning one or the other, but not both.

Reply:
For (3), without this case, then we would only be able to support return_types of ArrayType or StructType, which would remove support for scalar column outputs and require them to be represented as single-valued arrays, e.g. [1], [2].  I think this is entirely doable, but is less user-friendly.  Maybe a check of `len(preds) == batch_size` would suffice ?

Reply:
Added some [validation against the return types](https://github.com/apache/spark/pull/37734/commits/ca995dded8f1c722158ad73507edf098c74c80d3).

Reply:
> For the UDF return type, I tried using Iterator[Union[pd.Series, pd.DataFrame]], but got this error from pyspark/sql/pandas/typehints.py:135 during testing:
NotImplementedError: Unsupported signature: (data: 'Iterator[Union[pd.Series, pd.DataFrame]]') -> 'Iterator[Union[pd.Series, pd.DataFrame]]'.

This should be spark UDF bug @HyukjinKwon , we can ignore it for now and add a TODO here to fix it in future.

Reply:
> For (3), without this case, then we would only be able to support return_types of ArrayType or StructType, which would remove support for scalar column outputs and require them to be represented as single-valued arrays

Oh my fault, you are right, we should support return value to be (1) struct type (2) array type (3) scalar type.